include/deal.II/matrix_free/fe_evaluation.h

```diff
@@ -5778,17 +6064,87 @@ ::check_template_arguments(const unsigned int dof_no,
               ExcMessage(message));
     }
   if (dof_no != numbers::invalid_unsigned_int)
-    {
-      AssertDimension (n_q_points,
-                       this->mapping_data->descriptor[this->active_quad_index].n_q_points);
-      AssertDimension (this->data->dofs_per_component_on_cell * this->n_fe_components,
-                       this->dof_info->dofs_per_cell[this->active_fe_index]);
-    }
+    AssertDimension (n_q_points,
+                     this->mapping_data->descriptor[this->active_quad_index].n_q_points);
+#endif
+}
+
+
+
+template <int dim, int fe_degree,  int n_q_points_1d, int n_components_,
+          typename Number>
+inline
+void
+FEEvaluation<dim,fe_degree,n_q_points_1d,n_components_,Number>
+::reinit (const unsigned int cell_index)
+{
+  Assert (this->mapped_geometry == nullptr,
+          ExcMessage("FEEvaluation was initialized without a matrix-free object."
+                     " Integer indexing is not possible"));
+  if (this->mapped_geometry != nullptr)
+    return;
+
+  Assert (this->dof_info != nullptr, ExcNotInitialized());
+  Assert (this->mapping_data != nullptr, ExcNotInitialized());
+  this->cell = cell_index;
+  this->cell_type = this->matrix_info->get_mapping_info().get_cell_type(cell_index);
+
+  const unsigned int offsets = this->mapping_data->data_index_offsets[cell_index];
+  this->jacobian  = &this->mapping_data->jacobians[0][offsets];
+  this->J_value   = &this->mapping_data->JxW_values[offsets];
+
+#ifdef DEBUG
```

Comment:
why is it debug only?

Reply:
We only use these variables to track the state of the object in terms of assertions. We do not want to pay the price of initializing a variable that is never used in release mode. Note that this code has just been moved from `FEEvaluationBase` to `FEEvaluation`, in order to allow for `FEFaceEvaluation` to provide its own `reinit` without potential ambiguity from a `reinit` in a base class that is not appropriate.

Reply:
so declaration of those variables are also guarded in `#ifdef DEBUG... #endif`?